includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php

```diff
@@ -258,29 +277,125 @@ public function sanitize() {
 		 * This loop traverses through the DOM tree iteratively.
 		 */
 		while ( ! empty( $this->stack ) ) {
+			// Make sure we're targeting the first node of the remaining stack.
+			reset( $this->stack );
+			$node = current( $this->stack );
+			$node_index = key( $this->stack );
 
-			// Get the next node to process.
-			$node = array_shift( $this->stack );
-
-			/**
-			 * Process this node.
-			 */
+			// Process this node.
 			$this->process_node( $node );
 
+			// No need to check for children on contextual nodes.
+			if ( $node instanceof AMP_Contextual_Node ) {
+				array_shift( $this->stack );
+				continue;
+			}
+
+			// No need to proceed further if the node has no children.
+			if ( ! $node->parentNode ) {
+				array_shift( $this->stack );
+				continue;
+			}
+
 			/*
 			 * Push child nodes onto the stack, if any exist.
 			 * if node was removed, then it's parentNode value is null.
 			 */
 			if ( $node->parentNode ) {
-				$child = $node->firstChild;
-				while ( $child ) {
-					$this->stack[] = $child;
-					$child         = $child->nextSibling;
+
+				$child_nodes = $this->has_child_nodes( $node );
+
+				// If we need contextual nodes to track state for the current element,
+				// we add them as wrapping nodes around all children.
+				$contextual_nodes = $this->needs_contextual_nodes( $node );
+
+				if ( ! empty( $contextual_nodes ) ) {
+					array_unshift( $child_nodes, array_shift( $contextual_nodes ) );
+					array_push( $child_nodes, array_shift( $contextual_nodes ) );
 				}
+
+				// Replace the parent node by its child nodes.
+				array_splice( $this->stack, $node_index, 1, $child_nodes );
 			}
 		}
 	}
 
+	/**
+	 * Check whether a given node requires keeping track of context.
+	 *
+	 * If this is the case, it will return a pair of start and end nodes to
+	 * keep track of said context. This is done because we linearize the stack
+	 * and lack hierarchical information when processing the nodes. So we turn
+	 * one parent node into a pair of starting and ending contextual nodes.
+	 *
+	 * @param DOMNode $node Node to examine.
+	 * @return DOMNode[]|false Pair of contextual nodes, or false if none to be
+	 *                         used.
+	 */
+	protected function needs_contextual_nodes( DOMNode $node ) {
+		switch ( $node->nodeName ) {
```

Comment:
This `switch` only has one case so far, but this properly communicates the intent that there are multiple possible values we need to consider here. Turning it into an `if` for the short-term would obfuscate this intent and probably lead to problems later on when extending.